# :dna::dna::dna:    **M    u    t    a    P    i    p    e**     :dna::dna::dna:   
## A bioinformatics pipeline to identify high quality mutant and WT PDB structures for genes of interest
```diff
+NOTE: We are always working to improve MutaPipe, so any bug reports or suggestions are highly welcome.
```
## Table of Contents
1. [Introduction](#introduction)
2. [Citation](#citation)
3. [Documentation](#documentation)
  	* [Workflow](#workflow)
  	* [Incorporated Scripts](#incorporated-scripts)
	* [Minimum Requirements](#minimum-requirements)
	* [Local Deployment](#local-deployment)
	* [Usage](#usage)
	  * [Using the bash script](#using-the-bash-script)
	  	* [Usage example bash](#usage-example-bash)
	  * [Using the python scripts](#using-the-python-scripts)
	  	* [Usage example python](#usage-example-python)
	* [Output](#output)
	* [How to Download the Reference Proteome](#how-to-download-the-reference-proteome)

## Introduction

MutaPipe is a fast and efficient bioinformatics pipeline to screen the Protein Data Bank (PDB) for genes of interest and retrieve the highest quality protein structure (best resolution) for each unique sequence associated with a gene (currently only available for the species Homo sapiens). Additionally, whenever corresponding data is available, variants will be annotated using information on variant pathogenicity from ClinVar.

This allows researchers to efficiently screen the PDB for the most suitable template structures for a specific WT or mutant gene/protein which can be used for further in silico analysis like mutagenesis experiments or molecular dynmamics simulations.

![alt text](https://github.com/Utilon/Pipeline_Git/blob/main/files_for_README/Figure1_pipeline_simple_flowchart.jpeg)

Figure 1. *Pipeline overview. MutaPipe requires a list of gene names as input. The pipeline firstly performs a search of the entire Protein Data Bank (PDB) to retrieve information on all protein structures available in the PDB which are associated with the input genes (and the species: Homo sapiens). In a second step, a BLASTp against the canonical reference sequence for a given input gene (obtained from Uniprot) is performed for each sequence of every identified protein structure associated with said gene of interest. The BLASTp alignment is used to identify mismatches in the amino acid sequence of the PDB structures relative to the canonical sequence. In a subsequent step, the data is filtered and rearranged to output a table listing the highest quality structure for all available sequences in the PDB which are associated with the input genes. Additionally, if a given amino acid change in an input gene is listed in ClinVar, this information will be added to the output table.*



## Citation

MutaPipe has been developed by Deborah Ness, an NIHR Maudsley PhD student, in collaboration with and supervised by Dr Alfredo Iacoangeli, for her PhD project at the Department of Basic and Clinical Neuroscience at King's College London.

This work has not yet been published (Dec 2021). 

However, a dataset produced with Mutapipe for all genes available in the ClinVar Database has been presented as a poster at an international research conference in Italy in September 2021 ("THE 1Â° MASBIC - DISVA ANNUAL SYMPOSIUM - From structure to function: unveiling the role of proteins in health and disease")
The dataset is available [here](https://drive.google.com/drive/folders/1hqvAOLGMJuYqKuB1poH5_JPXFHDxwX80?usp=sharing).

## Documentation

### Workflow 

MutaPipe uses the inputted genes names (also possible in .txt format) to find all associated PDB IDs (using the PDB API) and download all corresponding fasta, mmCIF, and PDB files for further anaylses (will need storage depending on how many structures are available for the input genes (and their size)).
In a next step, MutaPipe parses all the downloaded files and combines all relevant information (e.g. resolution, sequence, unsolved residues in PDB structures). Subsequently, MutaPipe accessess the canonical protein sequence for all input genes (from Uniprot fasta file) and uses BLASTp (e-value threshold: .05) to identify all variants available in the identified PDB structures. BLASTp is performed for every sequence in every structure against the Uniport reference sequence for the gene associated with this structure (e.g. BLASTp of all sequences in a given FUS structure against the canonical sequence for FUS). The .xml output files generated by BLASTp are parsed and used to extract information, incl.: e-value, bit, alignment length, mismatches, close mismatches, gaps, indels.  This information will be used by MutaPipe identify all unique sequences in the PDB which are associated with the input genes and subsequently get the best structure (the one with the highest resolution) for all for each of them. 

MutaPipe achieves this by sorting the data to get:
- the best structure per point mutation (structures with only this one mutation and no other mutations)
- the best structure per unique combination of mutations available in the PDB (including WT/no mutation)
- the best structure for any specific mutation, regardless of other mutations in the same structure

Finally, whenever corresponding data is available, MutaPipe will annotate variants using variant information from ClinVar, incl. information on variant pathogenicity.

**Filtering**
To ensure sequences not associated with the gene of interest which are present in a PDB structure (e.g. if a protein of interest is co-crystallised with another peptide) do not get included in the final output, MutaPipe also has an additional step to filter out sequences:
1. which are shorter than a given percentage of the reference sequence (set variable relative_sequence_length; default value = 0.5)
2. whose best hsp covers less than a given percentage of the reference sequence (set variable hsp_coverage; default value = 0.1)




![alt text](https://github.com/Utilon/Pipeline_Git/blob/main/files_for_README/Figure2_Pipeline_detailled_workflow.jpeg)

Figure 2. *Detailled description of the steps performed by MutaPipe to identify variant and wildtype PDB structures associated with the input genes and identify the highest quality structure for each available sequence. (Note: The final step which adds available variant annotations from ClinVar is not displayed).*


### Incorporated Scripts

There are currently 9 different python scripts which are incorporated in MutaPipe:

| Script                                                                                                                        | Input                                                                                                                                                                                   | Operations                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | Output                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
|-------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| [00](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/00_search_pdb.py)                                              | genes in .txt format<br>(e.g. SOD1 ALS2 FUS)                                                                                                                                            | - creates a folder called Results in the current <br>working directory where all the output is going to be stored<br><br>- searches the pdb for all structures associated with <br>each gene name (in Homo Sapiens)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | - directory: Results/<br><br>- `00_search_overview_PDBids.csv`<br>*contains all gene names and corresponding PDB IDs if available*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| [01](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/01_download_files.py)                                          | `00_search_overview_PDBids.csv`                                                                                                                                                         | - creates a folder for each gene in the Results directory<br><br>- downloads mmCIF, pdb, and fasta files for all structures into the respective gene folder                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | - directory per gene: Results/GENENAME/<br><br>In each respective gene folder:<br>- mmCIF, pdb, and fasta files<br>stored in format `pdbID.cif/pdb/fasta`<br><br>In the Results folder:<br>- `01_search_overview_folders.csv`<br>*lists all the the newly created folders and their contents*<br>- `01_search_overview_n_structures.csv`<br>*lists number of structures retrieved per gene*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| [02](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/02_parse_cif_files.py)                                         | `01_search_overview_folders.csv`                                                                                                                                                        | **loops over all created folders and parses all mmCIf files, extracting:**<br><br>- resolution (999 for missing values/NMR structures)<br>- polypeptide sequences (corresponds to sequences as shown in PyMOL)<br>- fasta sequences                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | In each respective gene folder:<br>- `_ex.fasta` file for every structure <br>*fasta file created from the mmCIF file*<br>- `GENENAME_02_resolutions.csv`<br>*contains the resolution for all structures associated with this gene*<br>- `GENENAME_02_poly_seq.csv`<br>*contains all polypeptide sequences for all structures associated with this gene*<br><br>In the Results folder:<br>- `02_all_resolutions.csv`<br>*contains the resolutions of all parsed structures for all genes*<br>- `02_all_poly_seq.csv`<br>*contains all polypeptide sequences for all structures of all genes*                                                                                                                                                                                                                                                                                                                                                                 |
| [03](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/03_parse_fasta_files.py)                                       | `01_search_overview_folders.csv`                                                                                                                                                        | **loops over folders containing fasta and _ex.fasta files and:**<br><br>extracts info from fasta files, incl:<br>   - chain name<br>   - description<br>   - species<br>   - sequence<br><br>extracts info from _ex.fasta files, incl:<br>   - chain name<br>   - description<br>   - uniprot id <br>   - sequence<br><br>combines info from the two files                                                                                                                                                                                                                                                                                                                                                                                                                                                         | In each respective gene folder:<br>- `GENENAME_03_fasta_info.csv`<br>*contains information extracted from all fasta files for this gene*<br>- `GENENAME_03_fasta_ex_info.csv`<br>*contains information extracted from all _ex.fasta files for this gene*<br>- `GENENAME_03_fasta_combined_info.csv`<br>*contains combined information extracted from all fasta and _ex.fasta files for this gene*<br><br>In the Results folder: <br>- `03_fasta_info.csv`<br>*contains information extracted from all fasta files for all genes*<br>- `03_fasta_ex_info.csv`<br>*contains information extracted from all _ex.fasta files for all genes*<br>- `03_fasta_combined_info.csv`<br>*contains combined information extracted from all fasta and _ex.fasta files for all genes*                                                                                                                                                                                      |
| [04](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/04_blast_against_reference.py)                                 | `03_fasta_combined_info.csv`                                                                                                                                                            | - creates a directory called RefSeqs in the Results directory<br>............................................................................................................<br><br>- loops over input csv file (one row for each unique sequence in all the pdb structures for all genes)<br><br>- extracts the reference sequence for each gene from the Uniprot reference fasta into the RefSeqs directory<br><br>- writes a fasta file for each unique sequence/chain in the input csv into the RefSeqs directory<br><br>- performs BLASTp of all sequences against their respective reference sequence (e.g. FUS canonical sequence serves as reference for all sequences in all FUS structures) (output stored in .xml format in RefSeqs)<br><br>- uses the blast output (xml files) to identify mismatches | - directory: Results/Refseqs<br><br>In the RefSeqs folder:<br>- reference sequence fasta files<br>in format `GENE_reference.fasta`<br>- unique sequences fasta files<br>in format `GENE_pdbID_Chains.fasta`<br>- BLASTp output files (.xml)<br>in format `GENE_pdbID_Chains.xml`<br><br>In the Results folder: <br>- `04_blast_two_sequences.csv`<br>*lists all the information in the input file and the corresponding BLASTp results*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| [05](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/05_pdb_extract_unsolved_res.py)                                | `01_search_overview_folders.csv`<br><br>`04_blast_two_sequences.csv`                                                                                                                    | - loops over folders and extracts info on missing residues / residues which have not been solved in the crystal structure from each pdb file<br><br>- combines information on unsolved residues with info from 06_blast_fasta.csv / 08_blast_two_sequences.csv                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | In the Results folder: <br>- `05_unsolved_residues_per_structure.csv`<br>*lists all unsolved residues in all structures for all genes (one row for each structure)*<br>- `05_unsolved_residues_per_chain.csv`<br>*lists all unsolved residues in all chains of all structures for all genes (one row for each chain)*<br>- `05_all_info.csv`<br>*contains all the information from 04_blast_two_sequences.csv plus on unsolved residues extracted from pdb files*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
| [06](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/06_best_structure_per_mutation.py)                             | `05_all_info.csv`<br><br>`02_all_resolutions.csv`                                                                                                                                       | - combines the two dfs (according to PDBid)<br><br>- filters out sequences shorter than a given percentage of the reference sequence (set variable `relative_sequence_length`)<br><br>- filters out sequences whose best hsp covers less than a given percentage of the reference sequence (set variable `hsp_coverage`)<br><br>- sort/filter the df to get:<br> 1. the best structure per point mutation (structures with only this one mutation and no other mutations)<br> 2. the best structure per unique combination of mutations available in the PDB<br> 3. the best structure for any specific mutation, regardless of other mutations in the same structure                                                                                                                                              | In each respective gene folder:<br>- `GENENAME_06_best_structure_per_point_mutation.csv`<br>*lists best structure for each point mutation (one mutation per structure) in this gene*<br>- `GENENAME_06_best_structure_all_unique_combinations.csv`<br>*lists best structure for all unique mismatch combinations for this gene*<br>- `GENENAME_06_best_structure_any_mutation.csv`<br>*lists best structure for any mismatch in this gene regardless of other mismatches in this structure*<br><br>In the Results folder: <br>- `06_best_structure_per_point_mutation.csv`<br>*lists best structure for each point mutation (one mutation per structure) in all genes*<br>- `06_best_structure_all_unique_combinations.csv`<br>*lists best structure for all unique mismatch combinations for all genes*<br>- `06_best_structure_any_mutation.csv`<br>*lists best structure for any mismatch for all genes regardless of other mismatches in this structure* |
| [07_a](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/07_a_ClinVar_Annotations_edirect_per_gene_download_files.py) | `00_search_overview_availability.csv`                                                                                                                                                   | - query ClinVar for information on each gene of interest and download xml files with ids for all variants using edirect<br><br>- parse xml files with ClinVar ids and download/retrieve xml data from ClinVar for all variant ids associated with a gene                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | - all xml files downloaded from ClinVar are stored in the newly created ClinVar_Annotations folder (a subfolder in the Results directory)<br><br>- `07_a_ClinVar_Annotations_genes_no_data_retrieved.txt`<br>*lists all genes for which no ClinVar annotations could be retrieved*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| [07_b](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/07_b_ClinVar_Annotations_edirect_per_gene_parse_files.py)    | `00_search_overview_availability.csv`                                                                                                                                                   | - parse xml files and create a df with ClinVar information for all variants for all input genes                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | - `07_b_ClinVar_Annotations.csv`<br>*lists ClinVar annotations for all variants in all input genes*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| [08](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/08_add_clinvar_annotations_to_best_structures.py)             | `06_best_structure_all_unique_combinations.csv`<br><br>`06_best_structure_any_mutation.csv`<br><br><br>`06_best_structure_per_point_mutation.csv`<br><br>`07_b_ClinVar_Annotations.csv` | - adds availalable ClinVar annotations to all three best_structure tables (csv-files)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | - `08_best_structure_all_unique_combinations.csv`<br>*lists best structure for all unique mismatch combinations for all genes (incl. ClinVar annotations)*<br>- `08_best_structure_any_mutation.csv)`<br>*lists best structure for any mismatch for all genes regardless of other mismatches in this structure(incl. ClinVar annotations*<br>- `08_best_structure_per_point_mutation.csv`<br>*lists best structure for each point mutation (one mutation per structure) in all genes (incl. ClinVar annotations)*                                                                                                                                                                                                                                                                                                                                                                                                                                            |


### Minimum Requirements

- python > 3.6.8 (?)
- RAM: ?? (works fine with 16GB)
- Space required by the installation: 36 MB
- Scratch space for usage: depends the number and and size of PDB structures associated with the input genes. To check how many structures are available for your genes of interest, run `python3 00_search_pdb.py -g GENES_OF_INTEREST`. This will generate console output as well as csv files both of which provide an overview of how many structures have been identified for your input genes.

### Local Deployment

To obtain MutaPipe please use git to download the most recent development tree:

```bash
git clone https://github.com/Utilon/MutaPipe_Repo.git
```
#### NOTE: Reference Proteome Required
In order to run MutaPipe a fasta file containing the reference proteome / canonical protein sequences for the organism in question (= Homo sapiens) will have to be downloaded from Uniprot and stored in a folder called `Uniprot_reference_seqs` in the `MutaPipe` directory (where the python scripts are stored). The filename should be `UP000005640_9606.fasta` for the script to work properly.
The latest version of this fasta file can be accessed via the [Uniprot ftp client](https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/).

**NOTE**: When installing MutaPipe a fasta file of the reference proteome for Homo sapiens will be included (version from December 2021). Please feel free to update to a newer version if one is available (we will also regularly update the reference proteome file on here, but this is currently not done automatically).


### Usage

MutaPipe can be run in two main ways (from within the `MutaPipe`directory):
1. Using the bash script `MutaPipe.sh`
2. Using the MutaPipe python scripts

#### Using the bash script

MutaPipe can be run using the bash script `MutaPipe.sh`. MutaPipe.sh runs all scripts incorporated in the pipeline one after another. If you wish to run any of the pipeline stages separately, please follow the instructions [below](#Using-the-python-scripts). 

Its basic use requires the following three options:
```diff
+IMPORTANT: all three arguments are required in this specific order AND have to be written in double quotes!!!
```

```bash

  -g, --genes				specify input genes, to pass a file use \"-g \$(cat filename)\"
  -rsl, --relative_sequence_length	filter out sequences shorter than a given percentage of the reference sequence (0.1-1.0)
  -cov, --hsp_coverage			filter out sequences whose best hsp covers less than a given percentage of the reference sequence (0.1-1.0)

```

*Note:* More options are available when running the MutaPipe scripts manually. For further instructions, use the command `python3 script_name -h` on any of the MutaPipe python scripts or refer to the section on using the python scripts [below](#using-the-python-scripts).
 
##### Usage example bash

To run MutaPipe for gene NEK1, and setting the paramaters to filter out
1. sequences covering less than 50% of the reference sequence and 
2. sequences whose best hsp covers less than 10% of the reference sequence, 
use the command:

```bash
./MutaPipe.sh "-g NEK1" "-rsl 0.5" "-cov 0.1"
```

To run MutaPipe with **more than one input gene**, e.g. for NEK1 and SOD1, and the same parameters as before, use the command:
```bash
./MutaPipe.sh "-g NEK1 SOD1" "-rsl 0.5" "-cov 0.1"
```

To run MutaPipe passing a file containing the genes of interest, e.g. this example file called [genes.txt](https://github.com/Utilon/MutaPipe_Repo/blob/main/MutaPipe/genes.txt), and the same parameters as before, use the command:
```bash
./MutaPipe.sh "-g $(cat genes.txt)" "-rsl 0.5" "-cov 0.1"
```


#### Using the python scripts 

MutaPipe can be run manually, by running the 9 python scripts one after another (script [00](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/00_search_pdb.py), [01](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/01_download_files.py), [02](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/02_parse_cif_files.py), [03](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/03_parse_fasta_files.py), [04](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/04_blast_against_reference.py), [05](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/05_pdb_extract_unsolved_res.py), [06](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/06_best_structure_per_mutation.py), [07_a](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/07_a_ClinVar_Annotations_edirect_per_gene_download_files.py), [07_b](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/07_b_ClinVar_Annotations_edirect_per_gene_parse_files.py), [08](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/08_add_clinvar_annotations_to_best_structures.py)). More options are available when running the MutaPipe python scripts manually compared to when [using the bash script `MutaPipe.sh`](#using-the-bash-script). Use the `-h` flag on any MutaPipe python script to see instructions and available arguments (e.g. use the command: `python3 script_name -h`).

**All arguments for the MutaPipe python scripts are OPTIONAL.**
They have preset default values, i.e. they can be run *without* setting any of the available flags. To view the default values set in each MutaPipe python script, use the command `python3 script_name -h`.

**Script dependencies:** Although the scripts can be run individually, they mostly require at least some output from previous scripts in the pipeline. The table below lists the script dependencies in more detail:

*\[Table to be added here\]*


The following options are available for all the MutaPipe python scripts:

```bash
-h, --help		show help message and exit
-l, --log		write console output to log file in current directory if set to True, default = False
-t, --target		specify target directory, default = current_working_directory
```

Moreover, additional arguments can be set during different stages of MutaPipe:
 ```bash
# additional arguments for script 00_search_pdb.py
-g, --genes		Specify genes for which to search pdb structures, default = ['OPTN', 'ERBB4', 'DCTN1']; to pass a file containing all genes use -g $(cat filename)
-o, --organism		Specify species for which to search pdb structures, default = Homo sapiens
-a, --all		Retrieve all (True) vs max. 10 pdb IDs per gene (False), default = True

# additional arguments for script 01_download_files.py
-f, --format		Specify file format to be downloaded. For mmCif files (.cif) use 'cif' ; for pdb files (.pdb) use 'pdb' ; for fasta files (.fasta) use 'fasta' ; default = cif pdb fasta

# additional arguments for script 02_parse_cif_files.py
-pp, --polypeptides	Specify whether to extract polypeptide sequence (True) or not (False), default = True

# additional arguments for script 04_blast_against_reference.py
-bp, --blastp_path		Specify the path to blastp on your system ; default = blastp       
-refseq, --reference_sequences	Specify path to uniprot reference fasta, default = MutaPipe_Repo/MutaPipe/Uniprot_reference_seqs/UP000005640_9606.fasta    

# additional arguments for script 06_best_structure_per_mutation.py
-rsl, --relative_sequence_length	filter out sequences shorter than a given percentage of the reference sequence, default = 0.5
-cov, --hsp_coverage			filter out sequences whose best hsp covers less than a given percentage of the reference sequence, default = 0.1
```

##### Usage example python

```diff
+NOTE: This section is incomplete / not yet adequate for MutaPipe.
```

To run MutaPipe for gene NEK1, and setting the paramaters to filter out
1. sequences covering less than 50% of the reference sequence and 
2. sequences whose best hsp covers less than 10% of the reference sequence, 
use the command:

```bash
./MutaPipe.sh "-g NEK1" "-rsl 0.5" "-cov 0.1"
```

To run MutaPipe with **more than one input gene**, e.g. for NEK1 and SOD1, and the same parameters as before, use the command:
```bash
./MutaPipe.sh "-g NEK1 SOD1" "-rsl 0.5" "-cov 0.1"
```

To run MutaPipe passing a file containing the genes of interest, e.g. this example file called [genes.txt](https://github.com/Utilon/MutaPipe_Repo/blob/main/MutaPipe/genes.txt), and the same parameters as before, use the command:
```bash
./MutaPipe.sh "-g $(cat genes.txt)" "-rsl 0.5" "-cov 0.1"
```


### Output

```diff
+NOTE: This section is incomplete / not yet adequate for MutaPipe.
```

DNAscan output tree:

```bash

./$out_dir-| # this is the folder given to DNAsca using the -out flag. It will contain the aligned sequecing data ($sample_name.bam) as well as some temporanery files
           |
           |-results # This will contain the output of the analyses. E.g. $sample_name_sorted.vcf.gz , $sample_name_SV.vcf.gz, virus_results.txt, etc  
           |
           |-reports # If any report flags is used, this folder will contain the reports. E.g. $sample_name_vcfstats.txt if the -calls_report flag is used
           |
           |-logs # Logs files are generated in this folder
 
```

### How to Download the Reference Proteome

MutaPipe will need a fasta file with all canonical protein sequences of the species of interest.

In order to run MutaPipe a fasta file containing the reference proteome / canonical protein sequences for the organism in question (= Homo sapiens) will have to be downloaded from Uniprot and stored in a folder called `Uniprot_reference_seqs` in the `MutaPipe` directory (where the python scripts are stored). The filename should be `UP000005640_9606.fasta` for the script to work properly.
The latest version of this fasta file can be accessed via the [Uniprot ftp client](https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/).


**NOTE**: When installing MutaPipe (see [Local Deployment](#local-deployment)) a fasta file of the reference proteome for Homo sapiens will be included (version from December 2021). Please feel free to update to a newer version if one is available (we will also regularly update the reference proteome file on here, but this is currently not done automatically).

