# :dna::dna::dna:    **M    u    t    a    P    i    p    e**     :dna::dna::dna:   
## A bioinformatics pipeline to identify high quality mutant and WT PDB structures for genes of interest
```diff
+NOTE: We are always working to improve MutaPipe, so any bug reports or suggestions are highly welcome.
```
## Table of Contents
1. [Introduction](#introduction)
2. [Citation](#citation)
3. [Documentation](#documentation)
  	* [Workflow](#workflow)
  	* [Incorporated Scripts](#incorporated-scripts)
	* [Minimum Requirements](#minimum-requirements)
	* [Local Deployment](#local-deployment)
	* [Usage](#usage)
	  * [Using the bash script](#using-the-bash-script)
	  	* [Usage example bash](#usage-example-bash)
	  * [Using the python scripts](#using-the-python-scripts)
	  	* [Usage example python](#usage-example-python)
	* [Output](#output)
		* [Output example](#ouput-example)
	* [How to Download the Reference Proteome](#how-to-download-the-reference-proteome)

## Introduction

MutaPipe is a fast and efficient bioinformatics pipeline to screen the [Protein Data Bank (PDB)](https://www.rcsb.org/) for genes of interest and retrieve the highest quality protein structure (best resolution) for each unique sequence associated with a gene. Additionally, whenever corresponding data is available, variants will be annotated using information on variant pathogenicity from [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/intro/).

This allows researchers to efficiently screen the PDB for the most suitable template structures for a specific WT or mutant gene/protein which can be used for further in silico analysis such as homology modeling, mutagenesis experiments or molecular dynamics simulations.

![alt text](https://github.com/Utilon/Pipeline_Git/blob/main/files_for_README/Figure1_pipeline_simple_flowchart.jpeg)

**Figure 1.** Pipeline overview. MutaPipe requires a list of gene names as input. The pipeline firstly performs a search of the entire Protein Data Bank (PDB) to retrieve information on all protein structures available in the PDB which are associated with the input genes (and the species: Homo sapiens). In a second step, a BLASTp against the canonical reference sequence for a given input gene (obtained from [UniProt](https://www.uniprot.org/)) is performed for each sequence of every identified protein structure associated with said gene of interest. The BLASTp alignment is used to identify mismatches in the amino acid sequence of the PDB structures relative to the canonical sequence. In a subsequent step, the data is filtered and rearranged to output a table listing the highest quality structure for all available sequences in the PDB which are associated with the input genes. Additionally, if a given amino acid change in an input gene is listed in ClinVar, this information will be added to the output table.



## Citation

MutaPipe has been developed by Deborah Ness, an NIHR Maudsley Biomedical Research Centre PhD Student, in collaboration with and supervised by [Dr Alfredo Iacoangeli](https://www.kcl.ac.uk/people/alfredo-iacoangeli), for her PhD project at the [Department of Basic and Clinical Neuroscience at King's College London](https://www.kcl.ac.uk/neuroscience/about/departments/basic-clinical-neuroscience).

This work has not yet been published (Dec 2021). 

However, a dataset produced with MutaPipe for all genes available in the ClinVar Database has been presented as a poster at an international research conference in Italy in September 2021 ("THE 1Â° MASBIC - DISVA ANNUAL SYMPOSIUM - From structure to function: unveiling the role of proteins in health and disease")
The dataset is available [here](https://drive.google.com/drive/folders/1hqvAOLGMJuYqKuB1poH5_JPXFHDxwX80?usp=sharing).

## Documentation

### Workflow 

MutaPipe uses the inputted gene names to find all associated PDB IDs (using the PDB API) and download all corresponding `fasta`, `mmCif`, and `pdb` files for further anaylses (will require disk space dependant on how many structures are available for the input genes (and their size)).
In a next step, MutaPipe parses all the downloaded files and combines all relevant information (e.g. resolution, sequence, unsolved residues in PDB structures). Subsequently, MutaPipe accessess the canonical protein sequences for all input genes (from the UniProt fasta file) and uses *BLASTp* (e-value threshold: .05) to identify all variants available in the identified PDB structures. *BLASTp* is performed for every sequence in every structure against the UniProt reference sequence for the gene associated with this structure (e.g. *BLASTp* of all sequences in a given *FUS* structure against the canonical sequence for *FUS*). The `.xml` output files generated by *BLASTp* are parsed and used to extract information, incl.: e-value, bit, alignment length, mismatches, close mismatches, gaps, indels.  This information will be used by MutaPipe to identify all unique sequences in the PDB which are associated with the input genes and subsequently get the best structure (the one with the highest resolution) for each of them. 

MutaPipe achieves this by sorting the data to get:
- the best structure per point mutation (structures with only this one mutation and no other mutations)
- the best structure per unique combination of mutations available in the PDB (including WT/no mutation)
- the best structure for any specific mutation, regardless of other mutations in the same structure

Finally, whenever corresponding data is available, MutaPipe will annotate variants using variant information from ClinVar, incl. information on variant pathogenicity.

**Filtering** 
<br />
To ensure sequences not associated with the gene of interest which are present in a PDB structure (e.g. if a protein of interest is co-crystallised with another peptide) do not get included in the final output, MutaPipe also has an additional step to filter out sequences:
1. which are shorter than a given percentage of the reference sequence (set variable relative_sequence_length; default value = 0.5)
2. whose best High Scoring Segment Pair (HSP) covers less than a given percentage of the reference sequence (set variable hsp_coverage; default value = 0.1) *(Note: A high-scoring segment pair, or HSP, is a subsegment of a pair of sequences, either nucleotide or amino acid, that share high level of similarity. The level of similarity between the sequences depends on the settings of the local or global alignment algorithm which generated them)*

![alt text](https://github.com/Utilon/Pipeline_Git/blob/main/files_for_README/Figure2_Pipeline_detailled_workflow.jpeg)

**Figure 2.** Detailled description of the steps performed by MutaPipe to identify variant and wildtype PDB structures associated with the input genes and identify the highest quality structure for each available sequence. *Note:* The final step which adds available variant annotations from ClinVar is not displayed.


### Incorporated Scripts

```diff
+NOTE: The table in this section is not yet completely up to date.
```

There are currently 9 different python scripts incorporated in MutaPipe:

| Script                                                                                                                        | Input                                                                                                                                                                                   | Operations                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | Output                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
|-------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| [00](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/00_search_pdb.py)                                              | genes in .txt format<br>(e.g. SOD1 ALS2 FUS)                                                                                                                                            | - creates a folder called Results in the current <br>working directory where all the output is going to be stored<br><br>- searches the pdb for all structures associated with <br>each gene name (in Homo Sapiens)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | - directory: Results/<br><br>- `00_search_overview_PDBids.csv`<br>*contains all gene names and corresponding PDB IDs if available*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| [01](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/01_download_files.py)                                          | `00_search_overview_PDBids.csv`                                                                                                                                                         | - creates a folder for each gene in the Results directory<br><br>- downloads mmCIF, pdb, and fasta files for all structures into the respective gene folder                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | - directory per gene: Results/GENENAME/<br><br>In each respective gene folder:<br>- mmCIF, pdb, and fasta files<br>stored in format `pdbID.cif/pdb/fasta`<br><br>In the Results folder:<br>- `01_search_overview_folders.csv`<br>*lists all the the newly created folders and their contents*<br>- `01_search_overview_n_structures.csv`<br>*lists number of structures retrieved per gene*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| [02](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/02_parse_cif_files.py)                                         | `01_search_overview_folders.csv`                                                                                                                                                        | **loops over all created folders and parses all mmCIf files, extracting:**<br><br>- resolution (999 for missing values/NMR structures)<br>- polypeptide sequences (corresponds to sequences as shown in PyMOL)<br>- fasta sequences                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | In each respective gene folder:<br>- `_ex.fasta` file for every structure <br>*fasta file created from the mmCIF file*<br>- `GENENAME_02_resolutions.csv`<br>*contains the resolution for all structures associated with this gene*<br>- `GENENAME_02_poly_seq.csv`<br>*contains all polypeptide sequences for all structures associated with this gene*<br><br>In the Results folder:<br>- `02_all_resolutions.csv`<br>*contains the resolutions of all parsed structures for all genes*<br>- `02_all_poly_seq.csv`<br>*contains all polypeptide sequences for all structures of all genes*                                                                                                                                                                                                                                                                                                                                                                 |
| [03](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/03_parse_fasta_files.py)                                       | `01_search_overview_folders.csv`                                                                                                                                                        | **loops over folders containing fasta and _ex.fasta files and:**<br><br>extracts info from fasta files, incl:<br>   - chain name<br>   - description<br>   - species<br>   - sequence<br><br>extracts info from _ex.fasta files, incl:<br>   - chain name<br>   - description<br>   - uniprot id <br>   - sequence<br><br>combines info from the two files                                                                                                                                                                                                                                                                                                                                                                                                                                                         | In each respective gene folder:<br>- `GENENAME_03_fasta_info.csv`<br>*contains information extracted from all fasta files for this gene*<br>- `GENENAME_03_fasta_ex_info.csv`<br>*contains information extracted from all _ex.fasta files for this gene*<br>- `GENENAME_03_fasta_combined_info.csv`<br>*contains combined information extracted from all fasta and _ex.fasta files for this gene*<br><br>In the Results folder: <br>- `03_fasta_info.csv`<br>*contains information extracted from all fasta files for all genes*<br>- `03_fasta_ex_info.csv`<br>*contains information extracted from all _ex.fasta files for all genes*<br>- `03_fasta_combined_info.csv`<br>*contains combined information extracted from all fasta and _ex.fasta files for all genes*                                                                                                                                                                                      |
| [04](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/04_blast_against_reference.py)                                 | `03_fasta_combined_info.csv`                                                                                                                                                            | - creates a directory called RefSeqs in the Results directory<br>............................................................................................................<br><br>- loops over input csv file (one row for each unique sequence in all the pdb structures for all genes)<br><br>- extracts the reference sequence for each gene from the UniProt reference fasta into the RefSeqs directory<br><br>- writes a fasta file for each unique sequence/chain in the input csv into the RefSeqs directory<br><br>- performs BLASTp of all sequences against their respective reference sequence (e.g. FUS canonical sequence serves as reference for all sequences in all FUS structures) (output stored in .xml format in RefSeqs)<br><br>- uses the blast output (xml files) to identify mismatches | - directory: Results/Refseqs<br><br>In the RefSeqs folder:<br>- reference sequence fasta files<br>in format `GENE_reference.fasta`<br>- unique sequences fasta files<br>in format `GENE_pdbID_Chains.fasta`<br>- BLASTp output files (.xml)<br>in format `GENE_pdbID_Chains.xml`<br><br>In the Results folder: <br>- `04_blast_two_sequences.csv`<br>*lists all the information in the input file and the corresponding BLASTp results*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| [05](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/05_pdb_extract_unsolved_res.py)                                | `01_search_overview_folders.csv`<br><br>`04_blast_two_sequences.csv`                                                                                                                    | - loops over folders and extracts info on missing residues / residues which have not been solved in the crystal structure from each pdb file<br><br>- combines information on unsolved residues with info from 06_blast_fasta.csv / 08_blast_two_sequences.csv                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | In the Results folder: <br>- `05_unsolved_residues_per_structure.csv`<br>*lists all unsolved residues in all structures for all genes (one row for each structure)*<br>- `05_unsolved_residues_per_chain.csv`<br>*lists all unsolved residues in all chains of all structures for all genes (one row for each chain)*<br>- `05_all_info.csv`<br>*contains all the information from 04_blast_two_sequences.csv plus on unsolved residues extracted from pdb files*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
| [06](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/06_best_structure_per_mutation.py)                             | `05_all_info.csv`<br><br>`02_all_resolutions.csv`                                                                                                                                       | - combines the two dfs (according to PDBid)<br><br>- filters out sequences shorter than a given percentage of the reference sequence (set variable `relative_sequence_length`)<br><br>- filters out sequences whose best HSP covers less than a given percentage of the reference sequence (set variable `hsp_coverage`)<br><br>- sort/filter the df to get:<br> 1. the best structure per point mutation (structures with only this one mutation and no other mutations)<br> 2. the best structure per unique combination of mutations available in the PDB<br> 3. the best structure for any specific mutation, regardless of other mutations in the same structure                                                                                                                                              | In each respective gene folder:<br>- `GENENAME_06_best_structure_per_point_mutation.csv`<br>*lists best structure for each point mutation (one mutation per structure) in this gene*<br>- `GENENAME_06_best_structure_all_unique_combinations.csv`<br>*lists best structure for all unique mismatch combinations for this gene*<br>- `GENENAME_06_best_structure_any_mutation.csv`<br>*lists best structure for any mismatch in this gene regardless of other mismatches in this structure*<br><br>In the Results folder: <br>- `06_best_structure_per_point_mutation.csv`<br>*lists best structure for each point mutation (one mutation per structure) in all genes*<br>- `06_best_structure_all_unique_combinations.csv`<br>*lists best structure for all unique mismatch combinations for all genes*<br>- `06_best_structure_any_mutation.csv`<br>*lists best structure for any mismatch for all genes regardless of other mismatches in this structure* |
| [07_a](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/07_a_ClinVar_Annotations_edirect_per_gene_download_files.py) | `00_search_overview_availability.csv`                                                                                                                                                   | - query ClinVar for information on each gene of interest and download xml files with ids for all variants using edirect<br><br>- parse xml files with ClinVar ids and download/retrieve xml data from ClinVar for all variant ids associated with a gene                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | - all xml files downloaded from ClinVar are stored in the newly created ClinVar_Annotations folder (a subfolder in the Results directory)<br><br>- `07_a_ClinVar_Annotations_genes_no_data_retrieved.txt`<br>*lists all genes for which no ClinVar annotations could be retrieved*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| [07_b](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/07_b_ClinVar_Annotations_edirect_per_gene_parse_files.py)    | `00_search_overview_availability.csv`                                                                                                                                                   | - parse xml files and create a df with ClinVar information for all variants for all input genes                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | - `07_b_ClinVar_Annotations.csv`<br>*lists ClinVar annotations for all variants in all input genes*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| [08](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/08_add_clinvar_annotations_to_best_structures.py)             | `06_best_structure_all_unique_combinations.csv`<br><br>`06_best_structure_any_mutation.csv`<br><br><br>`06_best_structure_per_point_mutation.csv`<br><br>`07_b_ClinVar_Annotations.csv` | - adds availalable ClinVar annotations to all three best_structure tables (csv-files)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | - `08_best_structure_all_unique_combinations.csv`<br>*lists best structure for all unique mismatch combinations for all genes (incl. ClinVar annotations)*<br>- `08_best_structure_any_mutation.csv)`<br>*lists best structure for any mismatch for all genes regardless of other mismatches in this structure(incl. ClinVar annotations*<br>- `08_best_structure_per_point_mutation.csv`<br>*lists best structure for each point mutation (one mutation per structure) in all genes (incl. ClinVar annotations)*                                                                                                                                                                                                                                                                                                                                                                                                                                            |


### Minimum Requirements

- python > 3.6.8 (?)
- RAM: ?? (works fine with 16GB)
- Space required by the installation: 36 MB
- Scratch space for usage: depends the number and and size of PDB structures associated with the input genes. To check how many structures are available for your genes of interest, run the command `python3 00_search_pdb.py -g GENES_OF_INTEREST`. This will generate console output as well as csv files both of which provide an overview of how many structures have been identified for your input genes.

### Local Deployment

To obtain MutaPipe please use git to download the most recent development tree:

```bash
git clone https://github.com/Utilon/MutaPipe_Repo.git
```
#### NOTE: Reference Proteome Required
In order to run MutaPipe a fasta file containing the reference proteome / canonical protein sequences for the organism in question (= Homo sapiens) will have to be downloaded from UniProt and stored in a folder called `Uniprot_reference_seqs` in the `MutaPipe` directory (where the python scripts are stored). The filename should be `UP000005640_9606.fasta` for the script to work properly.
The latest version of this fasta file can be accessed via the [UniProt ftp client](https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/).

**NOTE**: When installing MutaPipe a fasta file of the reference proteome for Homo sapiens will be included (version from December 2021). Please feel free to update to a newer version if one is available (we will also regularly update the reference proteome file on here, but this is currently not done automatically).


### Usage

MutaPipe can be run in two main ways (from within the `MutaPipe`directory):
1. Using the bash script `MutaPipe.sh`
2. Using the MutaPipe python scripts

#### Using the bash script

MutaPipe can be run using the bash script `MutaPipe.sh`. MutaPipe.sh runs all scripts incorporated in the pipeline one after another. (If instead you wish to run any of the pipeline stages separately, please follow the instructions further [below](#Using-the-python-scripts)). 

Its basic use requires the following three options:

```bash

  Usage: [-h|l|t|g|o|a|f|p|b|u|r|c]
   
  General options:
  -h	HELP				Print this help message and exit
  -l	LOG				Write console output to log file in target directory if set to True. Default = False
  -t	TARGET_DIRECTORY  		Specify target directory where output will be stored. Default = current_working_directory
  
  MutaPipe options:
  -g	GENES				Specify genes of interest. To to pass a file containing all gene names use -g \$(cat filename). Default = default = ['OPTN', 'ERBB4', 'DCTN1']
  -o	ORGANISM			Set species for which to search pdb structures. Default = Homo sapiens
  -a	ALL_PDB_IDS		    	Specify whether to retrieve all (True) or max. 10 PDB IDs (False) per gene. Default = True
  -f	FORMAT			  	Specify file formats to download. Default = [cif pdb fasta]. Options = [cif pdb fasta]
  -p	POLYPEPTIDES			Specify whether to extract polypeptide sequence (True) or not (False). Default = True
  -b	BLASTp_PATH			Set path to blastp on your system. Default = blastp
  -u	UNIPROT_REFSEQS			Set path to reference proteome fasta file. Default = MutaPipe_Repo/MutaPipe/Uniprot_reference_seqs/
  -r	RELATIVE_SEQUENCE_LENGTH	Set to filter out sequences shorter than a given percentage of the reference sequence (0.1-1.0). Default = 0.5
  -c	HSP_COVERAGE			Set to filter out sequences whose best hsp covers less than a given percentage of the reference sequence (0.1-1.0). Default = 0.1   

```

##### Usage example bash

To run MutaPipe for gene NEK1 and set the paramaters to filter out
1. sequences covering less than 70% of the reference sequence and 
2. sequences whose best HSP covers less than 30% of the reference sequence, 
use the command:

```bash
./MutaPipe.sh -g NEK1 -r 0.7 -c 0.3
```

To run MutaPipe with **more than one input gene**, e.g. for NEK1 and SOD1, and the same parameters as before, use the command:
```bash
./MutaPipe.sh -g "NEK1 SOD1" -r 0.7 -c 0.3
```

To run MutaPipe passing a file containing the genes of interest, e.g. this example file called [genes.txt](https://github.com/Utilon/MutaPipe_Repo/blob/main/MutaPipe/genes.txt), and the same parameters as before, use the command:
```bash
./MutaPipe.sh -g $(cat genes.txt) -r 0.7 -c 0.3
```

#### Using the python scripts 

MutaPipe can be run manually, by running the 9 python scripts one after another (script [`00`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/00_search_pdb.py), [`01`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/01_download_files.py), [`02`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/02_parse_cif_files.py), [`03`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/03_parse_fasta_files.py), [`04`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/04_blast_against_reference.py), [`05`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/05_pdb_extract_unsolved_res.py), [`06`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/06_best_structure_per_mutation.py), [`07_a`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/07_a_ClinVar_Annotations_edirect_per_gene_download_files.py), [`07_b`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/07_b_ClinVar_Annotations_edirect_per_gene_parse_files.py), [`08`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/08_add_clinvar_annotations_to_best_structures.py)). Use the `-h` flag on any MutaPipe python script to see instructions and available arguments (e.g. use the command: `python3 script_name -h`).

**All arguments for the MutaPipe python scripts are OPTIONAL.** The scripts have preset default values, i.e. they can be run *without* setting any of the available flags. To view the default values set in each MutaPipe python script, use the command `python3 script_name -h`.

**Script dependencies:** Although the scripts can be run individually, they mostly require at least some output from previous scripts in the pipeline. The table below lists the script dependencies in more detail:

```diff
+Table to be added here
```

The following options are available for all the MutaPipe python scripts:

```bash
-h, --help		show help message and exit
-l, --log		write console output to log file in current directory if set to True, default = False
-t, --target		specify target directory, default = current_working_directory
```

Moreover, additional arguments can be set during different stages of MutaPipe:
 ```bash
# additional arguments for script 00_search_pdb.py
-g, --genes		Specify genes for which to search pdb structures, default = ['OPTN', 'ERBB4', 'DCTN1']; to pass a file containing all genes use -g $(cat filename)
-o, --organism		Specify species for which to search pdb structures, default = Homo sapiens
-a, --all		Retrieve all (True) vs max. 10 pdb IDs per gene (False), default = True

# additional arguments for script 01_download_files.py
-f, --format		Specify file format to be downloaded. For mmCif files (.cif) use 'cif' ; for pdb files (.pdb) use 'pdb' ; for fasta files (.fasta) use 'fasta' ; default = cif pdb fasta

# additional arguments for script 02_parse_cif_files.py
-pp, --polypeptides	Specify whether to extract polypeptide sequence (True) or not (False), default = True

# additional arguments for script 04_blast_against_reference.py
-bp, --blastp_path		Specify the path to blastp on your system ; default = blastp       
-refseq, --reference_sequences	Specify path to uniprot reference fasta, default = MutaPipe_Repo/MutaPipe/Uniprot_reference_seqs/UP000005640_9606.fasta    

# additional arguments for script 06_best_structure_per_mutation.py
-rsl, --relative_sequence_length	filter out sequences shorter than a given percentage of the reference sequence, default = 0.5
-cov, --hsp_coverage			filter out sequences whose best HSP covers less than a given percentage of the reference sequence, default = 0.1
```

##### Usage example python

Let's assume we want to run MutaPipe for the genes *SOD1*, *ERBB4*, and *DCTN1* using the python scripts one after another, so we have more direct control over the options at each stage of the pipeline.

Our genes of interest for this example are stored in a file called [`genes.txt`](https://github.com/Utilon/MutaPipe_Repo/blob/main/MutaPipe/genes.txt)

Using the first MutaPipe python script [`00_search_pdb.py`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/00_search_pdb.py) and the `-g` flag, we can search the PDB for structures associated with our genes of interest. Let's assume we also want to restrict our search to only retrieve max. 10 PDB ID's per input gene. We can do this by setting the `-a` flag to False.
```bash
python3 00_search_pdb.py -g $(cat genes.txt) -a False
```

The output files generated by `00_search_pdb.py` reveal that there are PDB structures for all our genes of interest. More specifically, our search retrieved:
1. 10 PDB IDs for *SOD1* (setting `-a` to True results in 127 PDB ID's)
2. 10 PDB IDs for *ERBB4*
3. 9 PDB IDs for *DCTN1*

Next, we use the scipt [`01_download_files.py`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/01_download_files.py) to download all relevant mmCif, pdb, and fasta files (for the PDB IDs identified in the previous step). While we could use the `-f`flag here to specify the desired file format to be downloaded, we don't set the flag as the default option (set to download mmCif, pdb, and fasta files) is required to run the full MutaPipe pipeline. We do not recommend changing the file format to be downloaded unless you don't want to run the rest of the pipeline - Remember that a complete MutaPipe run always requires downloading all three file formats (mmCif, pdb, fasta).
```bash
python3 01_download_files.py
```
Looking at our Results directory, we now see a new subfolder for each of our input genes (*SOD1*, *ERBB4*, and *DCTN1*) which contains all associated files retrieved from the PDB for this gene (mmCif, pdb, fasta). 

Now we are going to parse the downloaded mmCif files using the script [`02_parse_cif_files.py`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/02_parse_cif_files.py). We can set the flag `-pp` to True or False to indicate whether we want to also extract the polypeptide sequences from the mmCif files or not. The polypeptide sequences are continuous chains of amino acids without any missing residues in the crystal structure (comparable to how sequence is shown in PyMOL). It is not required to extract the polypeptide sequences for MutaPipe to run, so let's set the `-pp`flag to False:
```bash
python3 02_parse_cif_files.py -pp False
```
The output files generated by `02_parse_cif_files.py` list important information on all the structures, including structure name, resolution, structure method, deposition date, and classification.

Next, we are going to use the script [`03_parse_fasta_files.py`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/03_parse_fasta_files.py) to parse the downloaded fasta files as well as the fasta files created from the mmCif files (all in format \_ex.fasta) in the last step.
```bash
python3 03_parse_fasta_files.py
```
The output files generated by `python3 03_parse_fasta_files.py` list important information on all the sequences/chains in each fasta file, including chain name, UniProt ID, description, species, and amino acid sequence.

Now that all the files have been parsed, we can use the script [`04_blast_against_reference.py`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/04_blast_against_reference.py) to perform *BLASTp*. The script will perform a *BLASTp* for every sequence in every structure associated with our input genes against the corresponding reference sequence. By default, the reference sequence for each of our input genes will be retrieved from the UniProt fasta file containing the human reference proteome called [`UP000005640_9606.fasta`](https://github.com/Utilon/MutaPipe_Repo/blob/main/MutaPipe/Uniprot_reference_seqs/UP000005640_9606.fasta) which is included in MutaPipe (version December 2021; see instructions on how to download the latest version of this file [here](#How-to-Download-the-Reference-Proteome)). If we wanted to use a different file, we could set the `-refseq` flag and specify the path to the fasta file containing the reference proteome we'd like to use. 
```bash
python3 04_blast_against_reference.py
```
Looking at our Results directory, we can now see that a new subdirectory called `RefSeqs` has been created. It contains the canonical amino acid sequences for all of our input genes (*SOD1*, *ERBB4*, and *DCTN1*) as well as a subdirectory called `PDB_seqs_and_blastp_outputs` where all the PDB sequences used as query sequences in the *BLASTp* are stored in fasta format, and their corresponding blastp output files are stored in xml format. 

Moreover, the output files generated by script `04_blast_against_reference.py` nicely summarise all *BLASTp* results. Note that there are two files containing potential warnings: One in case no or more than one reference sequence is identified for an input gene and one in case *BLASTp* fails or does not produce any significant alignements. In our case 3 warnings for the gene *DCTN1* have been produced, namely:
1. No alignments for `DCTN1_2hl3_ChainC.fasta`
2. No alignments for `DCTN1_2hqh_ChainsE,F,G,H.fasta`
3. No alignments for `DCTN1_3e2u_ChainsE,F,G,H.fasta`
To investigate potential reasons for why there were no alignments for these sequences, we can have a look at these sequences in the output file called `04_blast_two_sequences.csv`. In our case, we can infer that the warnings have been generated because the sequences in question are actually not *DCTN1* sequences, but are other protein sequences which happen to be also present in/co-crystallised with a PDB structure associated with our input genes. The sequences which produced warnings in our case are not *DCTN1* sequences, but are instead 1. Microtubule-associated protein RP/EB family member 1, 2. Restin, and 3. CAP-Gly domain-containing linker protein 1.

In a next step, we use script [`05_pdb_extract_unsolved_res.py`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/05_pdb_extract_unsolved_res.py) to extract all unsolved residues in each crystall structure from the previously downloaded pdb files.
```bash
python3 05_pdb_extract_unsolved_res.py
```
The output files generated by `python3 05_pdb_extract_unsolved_res.py` provide an overview of all unsolved residues per structure and per chain. Unresolved/missing residues are amino acids which are absent in the crystall structure (no atomic coordinates available).

The next step in the pipeline consists of running the script [`06_best_structure_per_mutation.py`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/06_best_structure_per_mutation.py). It involves filtering out 
1. sequences which are too short (compared to the reference sequence) to be considered valuable template structures. We can use the flag `-rsl` to set this value (0.1-1.0).
2. sequences whose best High Scoring Segment Pair (HSP) covers less than a given percentage of the reference sequence We can use the flag `-cov` to set this value (0.1-1.0). *(Note: A high-scoring segment pair, or HSP, is a subsegment of a pair of sequences, either nucleotide or amino acid, that share high level of similarity. The level of similarity between the sequences depends on the settings of the local or global alignment algorithm which generated them)*
For our example we'll stick to the default settings which are `-rsl 0.5` and `-cov 0.1`, so we don't have to set the flags.
```bash
python3 06_best_structure_per_mutation.py
```
The 3 csv output files generated by `script 06_best_structure_per_mutation.py` now neatly lists the best PDB structure for each sequence available in the PDB which is associated with our input genes:
* The best structures for all sequences with only a single amino acid change (compared to canonical sequence) are listed in `06_best_structure_per_point_mutation.csv`
* The best structures for all unique sequences in the PDB associated with the input genes are listed in `06_best_structure_all_unique_combinations.csv`
* The best structures for any available variant in the PDB regardless of other variants in the same structure are listed in `06_best_structure_any_mutation.csv`


Finally we want to download the ClinVar annotations for all variants of our genes of interest so that we can add this information to the 3 csv files listing the best structures for all PDB sequences associated with the input genes whenever a ClinVar variant is available in a PDB structure.
We use the script [`07_a_ClinVar_Annotations_edirect_per_gene_download_files.py`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/07_a_ClinVar_Annotations_edirect_per_gene_download_files.py) which incorporates [Entrez Direct](https://www.ncbi.nlm.nih.gov/books/NBK179288/) to access ClinVar and download data on all variant ids associated with our input genes. Subsequently, we use script [`07_b_ClinVar_Annotations_edirect_per_gene_parse_files.py`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/07_b_ClinVar_Annotations_edirect_per_gene_parse_files.py) to parse the data downloaded from ClinVar. Finally, we add the relevant ClinVar annotations to the 3 csv files listing the best structures for all sequences associated with our input genes by running script [`08_add_clinvar_annotations_to_best_structures.py`](https://github.com/Utilon/Pipeline_Git/blob/main/MutaPipe/08_add_clinvar_annotations_to_best_structures.py).
```bash
python3 07_a_ClinVar_Annotations_edirect_per_gene_download_files.py
python3 07_b_ClinVar_Annotations_edirect_per_gene_parse_files.py
python3 08_add_clinvar_annotations_to_best_structures.py
```
Script `07_a_ClinVar_Annotations_genes_no_data_retrieved.txt` produces a file called `07_a_ClinVar_Annotations_genes_no_data_retrieved.txt`which contains a list of all genes for which no data has been retrieved from ClinVar. Script `07_b_ClinVar_Annotations_edirect_per_gene_parse_files.py` generates a table called `07_b_ClinVar_Annotations.csv` with information on all variants in ClinVar associated with our input genes. The final output files generated by script `08_add_clinvar_annotations_to_best_structures.py` contain all the information from the tables from before, including relevant ClinVar annotations whenever applicable:
* The best structures for all sequences with only a single amino acid change (compared to canonical sequence) are listed in `08_best_structure_per_point_mutation.csv`
* The best structures for all unique sequences in the PDB associated with the input genes are listed in `08_best_structure_all_unique_combinations.csv`
* The best structures for any available variant in the PDB regardless of other variants in the same structure are listed in `08_best_structure_any_mutation.csv`

### Output

MutaPipe creates a directory called `Results` where all intermediary and final main output files are stored (*Note:* when [running MutaPipe using python](#using-the-python-scripts), the `-t` flag lets you specify where you'd like this folder to be created, default = current working directory). All relevant files and data retrieved or extracted from the reference proteome fasta file, the ClinVar database, and the PDB are stored in additional subdirectories in the `Results` directory called `ClinVar_Annotations`, `RefSeqs`, and `[GENENAME]_[n]structures` (see [Figure 3](#ouput-example) below for an example).

Each of the nine python scripts incorporated in MutaPipe produces at least one main output file which will be stored directly in the `Results` directory. *NOTE:* All MutaPipe output filenames start with the numerical identifier of the corresponding MutaPipe script, making it easy to understand which outputs have been generated at which stage of the pipeline.

After a complete MutaPipe run with the default settings the following main output files will have been written to the `Results` directory:

```bash
- 00_search_overview_availability.csv 			* lists all input genes with a boolean value indicating whether corresponding PDB data exists
- 00_search_overview_PDBids.csv 			* lists all input genes with available PDB data and their corresponding PDB IDs
- 01_search_overview_folders.csv 			* lists all newly created directories (on for each input gene with available PDB data) and their content
- 01_search_overview_n_structures.csv 			* lists number of structures retrieved per gene (simple overview)
- 02_all_resolutions.csv 				* lists resolutions for all PDB structures associated with the input genes
- 02_all_poly_seq.csv 					* lists polypeptide sequences for all PDB structures associated with the input genes
- 03_fasta_info.csv 					* lists information extracted from all fasta files (downloaded from PDB) for all available sequences in the PDB associated with the input genes
- 03_fasta_ex_info.csv 					* lists information extracted from all fasta_ex files (extracted from mmCif files) for all available sequences in the PDB associated with the input genes
- 03_fasta_combined_info.csv 				* lists combined information extracted from all fasta and fasta_ex files for all available sequences in the PDB associated with the input genes
- 04_blast_two_sequences.csv 				* lists all the information in 03_fasta_combined_info.csv and the corresponding blastp results for every sequence
- 05_unsolved_residues_per_structure.csv 		* lists all unsolved residues in all PDB structures for all genes (one row for each structure)
- 05_unsolved_residues_per_chain.csv 			* lists all unsolved residues in all chains of all PDB structures for all genes (one row for each chain)
- 05_all_info.csv 					* lists all the information in 04_blast_two_sequences.csv and information on unsolved residues (extracted from PDB files)
- 06_best_structure_per_point_mutation.csv 		* lists the best structures for all sequences with only a single amino acid change (compared to canonical sequence)
- 06_best_structure_all_unique_combinations.csv 	* lists the best structures for all unique sequences in the PDB associated with the input genes (incl. WT/canonical sequence if corresponding structure is available)
- 06_best_structure_any_mutation.csv 			* lists the best structures for any available variant, i.e. amino acid change, in the PDB regardless of other variants in the same structure
- 07_a_ClinVar_Annotations_genes_no_data_retrieved.txt	* lists all genes for which no ClinVar annotations could be retrieved
- 07_b_ClinVar_Annotations.csv 				* lists all variants availalbe in ClinVar for the input genes
- 08_best_structure_all_unique_combinations.csv 	* lists the best structures for all unique sequences in the PDB associated with the input genes (incl. WT/canonical sequence if corresponding structure is available),  incl. relevant ClinVar annotations
- 08_best_structure_any_mutation.csv 			* lists the best structures for any available variant, i.e. amino acid change, in the PDB regardless of other variants in the same structure,  incl. relevant ClinVar annotations
- 08_best_structure_per_point_mutation.csv 		* lists the best structures for all sequences with only a single amino acid change (compared to canonical sequence),  incl. relevant ClinVar annotations
```

To avoid cluttering the directory containing the main output files, MutaPipe creates subfolders in the `Results` directory where it stores any additional files and relevant data retrieved or extracted from the reference proteome fasta file, the ClinVar database, and the PDB. 

*Note:* For each input gene with available PDB data, MutaPipe creates a subdirectory containing (1) data from the PDB retrieved via the PDB API (i.e. the corresponding mmCif, fasta, and pdb files) as well as (2) gene-specific output files generated at different steps in the pipeline. 

MutaPipe creates the following subdirectories in `Results`:

```bash
- ClinVar_Annotations 			* subdirectory where all ClinVar data is stored
- RefSeqs 				* subdirectory where all the reference sequences and other files used/generated by BLASTp are stored
- [GENENAME]_[n]structures 		* subdirectory for each input gene where all corresponding data from the PDB as well as gene-specific output files are stored
```

```diff
+NOTE: A section explaining all columns in main output files will be added shortly!
```

#### Ouput example

Let's assume we want to get the best PDB structures for all availbale NEK1-related sequences in the PDB. We stick to the default settings and thus use the following command to obtain results:
```bash
./MutaPipe.sh "-g NEK1" "-rsl 0.5" "-cov 0.1"
```

Consequently, we see that a new folder called `Results` has been created in our current working directory. It includes all the main output files as well as three additional subdirectories called `ClinVar_Annotations`, `RefSeqs`, and `NEK1_2structures`.  

![alt text](https://github.com/Utilon/MutaPipe_Repo/blob/main/files_for_README/Figure3_MutaPipe_Results.jpeg) 

**Figure 3.** Output generated by MutaPipe **a)** looking at the content in `Results` **b)** looking at the content in subdirectories `ClinVar_Annotations`, `RefSeqs`, and `NEK1_2structures`. *Note*: MutaPipe's main output files and gene-specific output files are highlighted in yellow in 3a. and 3b., respectively.


### How to Download the Reference Proteome

MutaPipe will need a fasta file with all canonical protein sequences of the species of interest.

In order to run MutaPipe a fasta file containing the reference proteome / canonical protein sequences for the organism in question (= Homo sapiens) will have to be downloaded from [UniProt](https://www.uniprot.org/) and stored in a folder called `Uniprot_reference_seqs` in the `MutaPipe` directory (where the python scripts are stored). The filename should be `UP000005640_9606.fasta` for the script to work properly.
The latest version of this fasta file can be accessed via the [UniProt ftp client](https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/).


**NOTE**: When installing MutaPipe (see [Local Deployment](#local-deployment)) a fasta file of the reference proteome for Homo sapiens will be included (version from December 2021). Please feel free to update to a newer version if one is available (we will also regularly update the reference proteome file on here, but this is currently not done automatically).

